### PR TITLE
ENH: implemented Owen's T function

### PR DESCRIPTION
**Pull request by evgenyzhurko**
#5160

### CODE REVIEW

File: scipy/special/basic.py [hunk 1]
```diff
@@ -32,7 +32,7 @@
            'keip_zeros', 'kelvin_zeros', 'ker_zeros', 'kerp_zeros', 'kv',
            'kvp', 'lmbda', 'lpmn', 'lpn', 'lqmn', 'lqn', 'mathieu_a',
            'mathieu_b', 'mathieu_even_coef', 'mathieu_odd_coef', 'ndtri',
-           'obl_cv_seq', 'pbdn_seq', 'pbdv_seq', 'pbvv_seq', 'perm',
+           'obl_cv_seq', 'owens_t','pbdn_seq', 'pbdv_seq', 'pbvv_seq', 'perm',
```
Comment (rgommers) [inline scipy/special/basic.py]: pep8, space after comma

File: scipy/special/basic.py [hunk 2]
```diff
@@ -2462,3 +2462,61 @@ def zeta(x, q=None, out=None):
...
     return _zeta(x, q, out)
 
+
+def owens_t(h, a):
+    r"""
+    Owen's T function.
+
+    Parameters
+    ----------
+    h: int or float
```
Comment (rgommers) [inline scipy/special/basic.py]: int or float --> scalar

(same 2 lines lower)

File: scipy/special/basic.py [hunk 3]
```diff
@@ -2462,3 +2462,61 @@ def zeta(x, q=None, out=None):
...
+    -------
+    t: float
+        Probability of the event (X > h and 0 < Y < a * X),
+        where X and Y are independent standard normal random variables.
+
+    References
+    ----------
+    .. [1] Mike Patefield and David Tandy. "Fast and accurate calculation
+           of Owen's T-function", 2000.
+           https://www.jstatsoft.org/article/view/v005i05/t.pdf
```
Comment (rgommers) [inline scipy/special/basic.py]: No link please, but a complete citation:

   M. Patefield and D. Tandy, "Fast and accurate calculation of Owen’st function",
   Statistical Software vol. 5, pp. 1-25, 2000.

File: scipy/special/basic.py [hunk 4]
```diff
@@ -2462,3 +2462,61 @@ def zeta(x, q=None, out=None):
...
+    if not (np.isscalar(h) and np.isscalar(a)):
+        raise ValueError("arguments must be scalars.")
+    if h < 0:
+        return owens_t(-h, a)
+    if a < 0:
+        return -owens_t(h, -a)
+    elif a > 1:
+        ncdf_h = ndtr(h)
+        ncdf_ah = ndtr(a * h)
+        return 0.5 * (ncdf_h + ncdf_ah) - ncdf_h * ncdf_ah -\
```
Comment (rgommers) [inline scipy/special/basic.py]: can you get rid of the ``\`` (replace by brackets)

File: scipy/special/basic.py [hunk 5]
```diff
@@ -2462,3 +2462,61 @@ def zeta(x, q=None, out=None):
...
+            owens_t(a * h, 1 / a)
+
+    if a == 0:
+        return 0
```
Comment (rgommers) [inline scipy/special/basic.py]: blank line here

File: scipy/special/basic.py [hunk 6]
```diff
@@ -2462,3 +2462,61 @@ def zeta(x, q=None, out=None):
...
+    if h == 0:
+        return 1 * np.arctan(a) / (2 * np.pi)
```
Comment (rgommers) [inline scipy/special/basic.py]: blank line here

File: scipy/special/tests/test_basic.py [hunk 7]
```diff
@@ -3416,5 +3416,33 @@ def xfunc(delta, r):
...
+
+    h = 100
+    a = None
+    assert_raises(ValueError, special.owens_t, h, a)
+
+    h = 3.2
+    a = 1.5
+    assert_almost_equal(special.owens_t(h, a), special.owens_t(h, np.inf))
+
+    delta = 0.5 * (cephes.ndtr(h) + cephes.ndtr(a * h)) -\
```
Comment (rgommers) [inline scipy/special/tests/test_basic.py]: same here, remove ``\``

File: scipy/special/basic.py [hunk 8]
```diff
@@ -2462,3 +2462,61 @@ def zeta(x, q=None, out=None):
...
+    z_prev = 0
+    z_curr = (ndtr(a * h) - 0.5) / h
+    ah_sq = -0.5 * h * h * a * a
+    pi_sq = np.power(2 * np.pi, -0.5)
+
+    for i in xrange(int(a * a * h * h / 2) + 1):
+        result += ((-1) ** (i - 1)) * z_curr
+        z_prev = z_curr
+        z_curr = ((2 * i - 1) * z_prev - pi_sq *
+            np.power(a, 2 * i - 1) * np.exp(ah_sq)) / (h * h)
```
Comment (rgommers) [inline scipy/special/basic.py]: blank line here

Comment (rgommers) [inline scipy/special/basic.py]: and style nitpick: lines like these should line up with the opening bracket above, not use 4-space indents

File: scipy/special/basic.py [hunk 9]
```diff
@@ -2462,3 +2462,61 @@ def zeta(x, q=None, out=None):
...
+        ncdf_h = ndtr(h)
+        ncdf_ah = ndtr(a * h)
+        return 0.5 * (ncdf_h + ncdf_ah) - ncdf_h * ncdf_ah -\
+            owens_t(a * h, 1 / a)
+
+    if a == 0:
+        return 0
+    if h == 0:
+        return 1 * np.arctan(a) / (2 * np.pi)
+    if a >= 0.99:
```
Comment (rgommers) [inline scipy/special/basic.py]: there's no test or comment on the PR for the accuracy when approaching 0.99 from either side. did you verify this?

Comment (rgommers) [inline scipy/special/basic.py]: and same question around ``1.0`` (other if clause above)

File: scipy/special/basic.py [hunk 10]
```diff
@@ -2462,3 +2462,61 @@ def zeta(x, q=None, out=None):
...
+    t: float
+        Probability of the event (X > h and 0 < Y < a * X),
+        where X and Y are independent standard normal random variables.
+
+    References
+    ----------
+    .. [1] Mike Patefield and David Tandy. "Fast and accurate calculation
+           of Owen's T-function", 2000.
+           https://www.jstatsoft.org/article/view/v005i05/t.pdf
+
```
Comment (rgommers) [inline scipy/special/basic.py]: maybe add an example which plots the function?

File: scipy/special/basic.py [hunk 11]
```diff
@@ -2462,3 +2462,61 @@ def zeta(x, q=None, out=None):
...
+    """
+    if not (np.isscalar(h) and np.isscalar(a)):
+        raise ValueError("arguments must be scalars.")
```
Comment (rgommers) [inline scipy/special/basic.py]: blank lines after an if-block ends (here and below)

**Comment by rgommers** [comment]
Thanks @evgenyzhurko. Would be good to mention that you want to add this function on the scipy-dev mailing list.

**Comment by pv** [comment]
Would make sense to implement this in Cython, to get a proper ufunc 
without restriction to scalar arguments.

**Comment by evgenyzhurko** [comment]
@rgommers I made a mistake wheni use `if a > 0.99` because this property can be used only for `a == 1`
Main part of algorithm can evaluate `T(h, a)` when `0 <= a <= 1 and h >= 0` and for this we use property: `T(h, a) = T(-h, a)`, `T(h, -a) = -T(h, a)`, `T(h, a) + T(h, 1/a) = 0.5 (F(h) + F(ah)) - F(h) * F(ah)`.

@pv I'll try to do it, but i have never use Cython before.

**Comment by ev-br** [comment]
@evgenyzhurko it's likely easiest to start from IPython's magic `%%cython -a```,  then move the code over to the `scipy/special/generate_ufuncs.py` framework.

**Comment by ewmoore** [comment]
You should import the tests for this function from Boost as well.  They have a table of values in https://github.com/boostorg/math/blob/develop/test/owens_t.ipp that can be added to the similar tests as well as a bunch of other tests that can be ported over.  No reason not to make use of work already done to construct some good tests.

**Comment by person142** [comment]
You reference M. Patefield and D. Tandy, but it seems that the algorithm used in the paper is much more complex. Is that true, and if so, how did you achieve the simplifications?

**Comment by pv** [comment]
It's the method T2 from Patefield&Tandy. Probably covers only part of 
the parameter region, and if so more work is necessary here.

**Comment by person142** [comment]
Just to second @pv, to make this production-ready it should be rewritten in Cython (or C, or C++, or Fortran) and hooked up to the ufuncs machinery.

File: scipy/special/add_newdocs.py [hunk 12]
```diff
@@ -6369,3 +6369,8 @@ def add_newdoc(place, name, doc):
     """
     Internal function, do not use.
     """)
+
+add_newdoc("scipy.special", "owens_t",
+    """
```
Comment (person142) [inline scipy/special/add_newdocs.py]: Instead of putting the documentation in `owens_t.c` put it here.

File: scipy/special/cephes/owens_t.c [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+static double WTS[] = {
+	0.18831438115323502887E-01, 0.18567086243977649478E-01,
+    0.18042093461223385584E-01, 0.17263829606398753364E-01,
+    0.16243219975989856730E-01, 0.14994592034116704829E-01,
+    0.13535474469662088392E-01, 0.11886351605820165233E-01,
+    0.10070377242777431897E-01, 0.81130545742299586629E-02,
+    0.60419009528470238773E-02, 0.38862217010742057883E-02,
+    0.16793031084546090448E-02};
+
+static double PI = 3.1415926535897932384626433832795;
```
Comment (person142) [inline scipy/special/cephes/owens_t.c]: Use `NPY_PI`.

File: scipy/special/cephes/owens_t.c [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+		if (h <= HRANGE[i]) {
+			ihint = i;
+			break;
+		}
+	}
+
+	return SELECT_METHOD[iaint* 15 + ihint];
+}
+
+int get_ord(int index) {
```
Comment (person142) [inline scipy/special/cephes/owens_t.c]: I don't think these helper functions are necessary.

File: scipy/special/cephes/owens_t.c [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+
+    double result = (0.5 * (ndtr(h)) * (1 - ndtr(h)) - atan((1 - a) /
+    	 (1 + a)) * exp(-0.5 * (1 - a) * h * h / atan((1 - a) /
+    	 (1 + a))) / (2 * PI));
+
+    return result;
+}
+
+double owens_t(double h, double a) {
+	if (cephes_isnan(h) || cephes_isnan(a))
```
Comment (person142) [inline scipy/special/cephes/owens_t.c]: Braces.

File: scipy/special/cephes/owens_t.c [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+             exp(ah_sq))) / (h_sq);
+    }
+
+    result *= -pi_sq * exp(-0.5 * h_sq);
+
+    return result;
+}
+
+double owensT3(double h, double a, double m) {
+    double a_sq = a * a;
```
Comment (person142) [inline scipy/special/cephes/owens_t.c]: Can you name them something like `a2` or `aa`?

File: scipy/special/cephes/owens_t.c [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+    double maxi = 2 * m + 1;
+    double h_sq = h * h;
+    double a_sq = -a * a;
+
+    int i = 1;
+    double ai = a * exp(-h_sq * (1 - a_sq) / 2) / (2 * PI);
+    double yi = 1;
+    double result = 0;
+
+    while (1) {
```
Comment (person142) [inline scipy/special/cephes/owens_t.c]: Why not `for (i = 1; i < maxi; i += 2)`?

File: scipy/special/cephes/owens_t.c [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+
+static double PI = 3.1415926535897932384626433832795;
+
+int get_method(double h, double a) {
+	int ihint = 14;
+	int iaint = 7;
+	int i = 0;
+
+	for (i = 0; i < 14; i++) {
+		if (h <= HRANGE[i]) {
```
Comment (person142) [inline scipy/special/cephes/owens_t.c]: Spacing isn't consistent with the rest of the code here.

**Review by person142 — COMMENTED** [review]
Looking good--the build is failing because you need to add a line to `test_cython_special.py`.

File: scipy/special/tests/test_basic.py [hunk 19]
```diff
@@ -3416,5 +3416,412 @@ def xfunc(delta, r):
     assert_func_equal(special.pseudo_huber, w, z, rtol=1e-13, atol=1e-13)
 
 
+owenst_dt = [[(1.9508080482482910156250000000000000000000000000000000000000000000000000000000000000000000000000000000e+00), (9.7540378570556640625000000000000000000000000000000000000000000000000000000000000000000000000000000000e-02), (2.2942365980155351117754771526378292708740231264960539316842063696200884829628237039895484898529192823e-03)],
```
Comment (pv) [inline scipy/special/tests/test_basic.py]: If this is the test data from boost, it should be put in a file under `data/boost/owenst_ipp` and added to the TESTS list in `test_data.py:test_boost`. (Also if it's some other test data, long data set like this should be put in a data file.)

After adding the text data, the npz data file needs to be regenerated using utils/makenpz.py

File: scipy/special/cephes/owens_t.c [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+			ihint = i;
+			break;
+		}
+	}
+
+	for (i = 0; i < 7; i++) {
+		if (h <= HRANGE[i]) {
+			ihint = i;
```
Comment (pv) [inline scipy/special/cephes/owens_t.c]: iahint?

**Comment by evgenyzhurko** [comment]
I add `(special.owens_t, cython_special.owens_t, ('dd',), None)` to `test_cython_special.py` but it's not solve the previous fail. I try to run test using command `python runtest.py -v` and tests stoped without error. 
End of logs:
`test_cython_special.test_cython_api((<ufunc 'zetac'>, <built-in function zetac>, ('d',), None),) ... ok`
`test_cython_special.test_cython_api((<ufunc 'owens_t'>, <built-in function owens_t>, ('dd',), None),) ...`

**Comment by person142** [comment]
Now there are different failures--looks like the docstring has some issues and at least one of the builds is getting a segfault. You can go look at the build logs for details.

**Comment by evgenyzhurko** [comment]
How i can correct issue with docstring?
`Failed example:`
   `special.owens_t(h, a)`
`Expected:`
   `0.10877216734852269`
`Got:`
   `0.10877216734852269`

**Comment by ev-br** [comment]
> Expected:
0.10877216734852269
Got:
0.10877216734852269

Something about locale / non-ascii symbols?

**Comment by evgenyzhurko** [comment]
There is no message about locale or ascii. I think it's a result of others error or segmentation foult, because in previous commit all is good with docstring.

**Comment by person142** [comment]
> in previous commit all is good with docstring

It wasn't really checked in the previous commit because most of the documentation was in the C source; the refguide checker only looks at the Python docstrings.

**Comment by evgenyzhurko** [comment]
Main mistake that i put all changes into 1 commit. I'll divide changes into several commit(move tests, change docstring) so it will be easier to find problems.

File: scipy/special/tests/test_data.py [hunk 21]
```diff
@@ -430,6 +430,8 @@ def test_boost():
         data(spherical_jn_, 'sph_bessel_data_ipp-sph_bessel_data', (0,1), 2, rtol=1e-13),
         data(spherical_yn_, 'sph_neumann_data_ipp-sph_neumann_data', (0,1), 2, rtol=8e-15),
 
+        data(owens_t, 'owenst_data_ipp-owenst', (0, 1), 2, rtol=5e-1),
```
Comment (person142) [inline scipy/special/tests/test_data.py]: If the required `rtol` is this large there's probably a bug in the code.

Comment (evgenyzhurko) [inline scipy/special/tests/test_data.py]: I found difference between between real result and result  of `owens_t` function. 
Part of results:
Sample number |  Algorithm | Difference
10 T4 -0.00645986404905
11 T4 -0.00674801958045
12 T4 -0.00772143934778
13 T4 -0.00782197442148
14 T4 -0.00838921922808
15 T4 -0.0084809175162
16 T4 -0.00851552649868
17 T4 -0.00852995038951
18 T4 -0.00881967499371
19 T4 -0.00886193796425
20 T4 1.08420217249e-19
21 T4 2.16840434497e-19
22 T4 -2.16840434497e-19
23 T4 2.16840434497e-19
24 T4 2.16840434497e-19
25 T4 2.16840434497e-19

One algorithm gives difference between real results and function's result from ~0.1e-2 to ~1e-20.

Possible solution: implement all additional algorithm from http://www.boost.org/doc/libs/1_52_0/boost/math/special_functions/owens_t.hpp

Comment (person142) [inline scipy/special/tests/test_data.py]: Are those relative errors or absolute errors? An absolute error of `1e-20` could still be bad.

Comment (evgenyzhurko) [inline scipy/special/tests/test_data.py]: Min relative error 1.3031421275295663e-16 with absolute error 1.0842021724855044e-19
Max relative error 3.9470692224712853 with absolute error 2.319836168418231e-08
There is also examples with relative and absolute error = 0

File: scipy/special/cephes/owens_t.c [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+    return result;
+}
+
+double owensT4(double h, double a, double m) {
+    double maxi = 2 * m + 1;
+    double hh = h * h;
+    double aa = -a * a;
+
+    int i = 1;
+    double ai = a * exp(-hh * (1 - aa) / 2) / (2 * NPY_PI);
```
Comment (person142) [inline scipy/special/cephes/owens_t.c]: Should be `exp(-hh * (1 + aa) / 2)`?

File: scipy/special/cephes/owens_t.c [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+    double yi = 1;
+    double result = 0;
+
+    while (1) {
+        result += ai * yi;
```
Comment (person142) [inline scipy/special/cephes/owens_t.c]: Only need to multiply by `ai` at the end of the loop?

File: scipy/special/tests/test_data.py (hunk 21 above)
Comment (person142) [inline scipy/special/tests/test_data.py]: I think there's a bug? See comment in `owensT4`.

File: scipy/special/cephes/owens_t.c (hunk 22 above)
Comment (evgenyzhurko) [inline scipy/special/cephes/owens_t.c]: No, it's right code. T4(h, a, m) = ((2Pi) ^ -1) *  exp(- 0.5* h*h*(1 + a*a))* Sum\[0 to m](((-a*a)^i)*yi)

File: scipy/special/cephes/owens_t.c (hunk 23 above)
Comment (evgenyzhurko) [inline scipy/special/cephes/owens_t.c]: No, we can myltiply in loop or after loop. Both variant of algorithms  give the same result.

Comment (person142) [inline scipy/special/cephes/owens_t.c]: Well yes, but multiplying after requires less multiplications...

File: scipy/special/cephes/owens_t.c (hunk 22 above)
Comment (person142) [inline scipy/special/cephes/owens_t.c]: Could you change `aa` to something like `naa`? (And similar for some other variables in the code.)

Comment (person142) [inline scipy/special/cephes/owens_t.c]: Though I guess this is following what Boost does, so maybe that's reasonable.

File: scipy/special/tests/test_data.py (hunk 21 above)
Comment (person142) [inline scipy/special/tests/test_data.py]: Ok, you've quite ably defended against my accusations of bugs :-), so let's see whether the error should be that large. For the point which has large relative error: take the sum of the absolute value of the terms of the series and divide it by the sum of the series to get the condition number. Multiply the condition number by machine epsilon; if it's about the same size as the relative error then it's a mathematical difficulty (cancellation in an alternating series), and it's time to implement the extra methods in Boost. If it's much smaller then there's a bug.

**Comment by person142** [comment]
Would be good to figure out how well Boost does on its own tests so that we know this implementation is on par.

**Comment by evgenyzhurko** [comment]
Boost only checks result within expected bounds and not specify correct tolerance.
https://github.com/boostorg/math/blob/242e9d669198415d165bb0213f730c00985bee9a/test/handle_test_result.hpp#L115

**Comment by person142** [comment]
Run the Boost function on the Boost data.

**Comment by person142** [comment]
Also will need more tests to cover the (at least) `20 < a < 200` region; should also test the negative `a`, `h` region but this can be done simply with a self-consistency check because of the symmetry.

**Comment by evgenyzhurko** [comment]
I copied last part of the test cases(`local/owens_t_tabullated.txt`) from http://people.sc.fsu.edu/~jburkardt/f77_src/owens/owens_prb_output.txt

**Comment by person142** [comment]
IIRC the stuff from there is usually GPL'ed--i.e. not sure if we can use that data (am actually not sure if licenses apply to data in the same way).

File: scipy/special/add_newdocs.py [hunk 24]
```diff
@@ -6369,3 +6369,36 @@ def add_newdoc(place, name, doc):
...
```
Comment (person142) [inline scipy/special/add_newdocs.py]: Begin the documentation with a one sentence definition; i.e. "Owen's T Function."

Then there's a blank line, followed by the main definition.

File: scipy/special/add_newdocs.py [hunk 25]
```diff
@@ -6369,3 +6369,36 @@ def add_newdoc(place, name, doc):
...
+    The function T(h, a) gives the probability of the event
+    (X > h and 0 < Y < a * X) where X and Y are independent
+    standard normal random variables.
+
+    Parameters
+    ----------
+    h: float64
```
Comment (person142) [inline scipy/special/add_newdocs.py]: array_like instead of float64.

File: scipy/special/add_newdocs.py [hunk 26]
```diff
@@ -6369,3 +6369,36 @@ def add_newdoc(place, name, doc):
...
+        Input value.
+    a: float64
+        Input value.
+ 
+    Returns
+    -------
+    t: float64
```
Comment (person142) [inline scipy/special/add_newdocs.py]: scalar or ndarray instead of float64.

File: scipy/special/tests/test_cython_special.py [hunk 27]
```diff
@@ -271,7 +271,8 @@ def test_cython_api():
         (special.yn, cython_special.yn, ('ld', 'dd'), None),
         (special.yv, cython_special.yv, ('dd', 'dD'), None),
         (special.yve, cython_special.yve, ('dd', 'dD'), None),
-        (special.zetac, cython_special.zetac, ('d',), None)
+        (special.zetac, cython_special.zetac, ('d',), None),
+        (special.owens_t, cython_special.owens_t, ('dd',), None)
```
Comment (person142) [inline scipy/special/tests/test_cython_special.py]: This big list was in alphabetical order; let's keep it that way.

**Comment by person142** [comment]
It might be safer to use `mpmath` to generate more test data. (And you could get more of it.)

**Comment by evgenyzhurko** [comment]
@person142 Last data is GNU LGPL licensed. If i correctly understood license I can include this data in Scipy. https://people.sc.fsu.edu/~jburkardt/m_src/owens/owen_values.m

**Comment by pv** [comment]
I've been told that the GPL/LGPL attributions on that page are not all 
accurate, so I would be careful using that as a resource.

**Comment by evgenyzhurko** [comment]
https://github.com/boostorg/math/blob/develop/example/owens_t_example.cpp
Boost contains this data also.

**Comment by evgenyzhurko** [comment]
Oh, no, it's other data, but that also have test case with a > 1, so i can use it.

**Comment by evgenyzhurko** [comment]
@person142 I tested Boost owens_t function on Boost data and get absolute error equals to 0 for all(400) test cases. It's very strange that absolute error equal to 0 even in value  ~5e-89.

**Comment by person142** [comment]
Like literal floating point equality equal to zero? That doesn't sounds right...

**Comment by evgenyzhurko** [comment]
I tested with double type and got 0. Now i test with long double and get  minimal absolute error equal to 1.7194768829262296883086946763796731829643249511719e-19

**Comment by person142** [comment]
Then something is going wrong because 1e-19 is very well within double precision range. Also, Owen's T gets small very quickly, which is why you need to be measuring relative errors.

**Comment by person142** [comment]
e.g. if the true value of the function is on the order of 1e-89 and you get an absolute error of 1e-19 then there's a problem. (Unless near a zero of the function.)

**Comment by person142** [comment]
Unless the Boost function is actually correctly rounded on all that test data, which would be impressive.

**Comment by evgenyzhurko** [comment]
Relative error for all test cases = from 1e-17 to 7e-18

**Comment by evgenyzhurko** [comment]
When i wrote aboute absolute error, i made a mistake, 1.7194768829262296883086946763796731829643249511719e-19 is maximal absolute error, minimal absolute error is about 7e-106 for data ~1e-89

**Comment by evgenyzhurko** [comment]
Relative tolerance for this data is from 1e-17 to 7e-19

**Comment by person142** [comment]
Is Boost using extended precision internally for the double precision routine? That would explain this. (Would just check myself but am away from computer.)

**Comment by evgenyzhurko** [comment]
I tested `owens_t `(from this pull request) with different tolerance and i got error only for 2 test case(`rtol=5e-14`), rdiff in this 2 test case ~1e-8, others test case have rdiff about 5e-14 and less. 2 test case with problem use T6 algorithm for computing.

If I use T5 algorithm for 2 problem test case, tests passed with rtol = 5e-14.

**Comment by evgenyzhurko** [comment]
I found a bug in my code. Current `owens_t` function passes tests with `rtol=5e-14`

**Comment by person142** [comment]
Very nice!

**Comment by evgenyzhurko** [comment]
@person142 Do you expect more accuracy in computing for this pull request or some other changes?

**Comment by person142** [comment]
`5e-14` is certainly sufficient. I still need to review the mathematics in detail, but that's on me.

Probably the most important thing now is using `gcov` to make sure all of the different code branches are getting executed and adding additional tests if some of them aren't. (Since the selection routine for the function is so complicated it will be difficult to feel confident that all the cases are being tested otherwise.)

**Comment by evgenyzhurko** [comment]
Thanks, i'll try to generate test case for all code branches.

**Comment by evgenyzhurko** [comment]
Boost test cases covers `T3`, `T5`, `T6`, `T4_mp` algorithms.
I added test cases for `T1`, `T2`, `T1 accelerated` and `T2 accelerated`, but i didn't find test case for `T4`. 
I found the whole range of values `a` and `h` for `T4` method and tested what is method used to compute for this value, as a result all this values use `T4_mp` algorithm. I used step 0.0001 and can't more(it takes very long time).
I have no test case fot `T4` algorithm, but i think it's not necessary delete `T4` algorithm, because `T4_mp` isn't stable in some cases(but i didn't find this cases for `T4` algorithm).

File: scipy/special/add_newdocs.py (hunk 24 above)
Comment (pv) [inline scipy/special/add_newdocs.py]: The first line should actually be the function signature, i.e.,
```
owens_t(h, a)

Owen's T Function.

...
```

**Comment by evgenyzhurko** [comment]
@person142 gcov shows 92.5% coverage for lines and 93.3% for functions. As I say only OwensT4 function not coveraged yet, because `OwenstT4_mp` covers range of `OwensT4`, but `OwenstT4_mp` might be unstable in some case.
image
image

**Comment by ev-br** [comment]
Josh, @person142, any chance you can take a look?

**Comment by person142** [comment]
Yes, sorry @evgenyzhurko for sitting on this for so long... will try to take a final look; probably won't happen until next week though.

**Comment by xoviat** [comment]
@evgenyzhurko Can you rebase on `master`?

**Comment by evgenyzhurko** [comment]
I rebased this PR on master, fixed all build problem and merge conflicts. Is anyone can review this PR?

**Comment by person142** [comment]
@evgenyzhurko I rebased again (recent changes in `cephes` conflicted) and made some fixes. Still need to investigate the test coverage a little more.

Probably most importantly: from reading the code it looks like the Boost implementation was followed closely, so I added the Boost license information to the top of the file. Am I correct?

**Comment by evgenyzhurko** [comment]
@person142 You are right, in many case i follow to boost implementation.

I was trying to find values a and h that can be calculated using only T4(that not covered yet) function but it's no use. I calculated that for 0.06 < a < 0.25 and 0.6 < h < 3.5 should be used T4 function. I iterated through this interval with step=0.00001 and T4 wasn't used.

**Comment by person142** [comment]
> I was trying to find values a and h that can be calculated using only T4(that not covered yet) function but it's no use

It appears you were using Boost's arbitrary precision dispatch function instead of the double precision version. I switched to using the double precision version (which covers T4), fixed a bug this revealed in T1, and then removed the no-longer-needed arbitrary precision helper functions.

**Comment by evgenyzhurko** [comment]
@person142 Thanks for your help and my mistakes explanation. Is something else need from me in this PR?

**Comment by person142** [comment]
> Is something else need from me in this PR?

I made a few more fixes, and I think this is good to go now.

We will see what the CIs say, and we should probably ping @pv to do a sanity check since I feel a little too involved in the code editing process to be an objective third party. Pauli, I'm seeing 100% test coverage on `owens_t.c` with the exception of the `default` case on line 308, which should in theory never be reached (it's just a sentinel).

**Comment by person142** [comment]
Ah shoot, looks like `cephes/protos.h` and some submodule updates got picked up in the rebase--will have to fix that.

**Comment by person142** [comment]
Ok, cleaned up the history. Now there's one clean commit.

**Comment by person142** [comment]
Ok, got this in at long last. Thanks for your patience @evgenyzhurko.

**Pull request merged**